File: 49.字母异位词分组.go
/*
 * @lc app=leetcode.cn id=49 lang=golang
 *
 * [49] 字母异位词分组
 */

// @lc code=start
func groupAnagrams(strs []string) [][]string {
	tmp := make(map[[127]int][]string)
	for _, str := range strs {
		cnt := [127]int{0}
		for _, v := range str {
			cnt[v-'a']++
		}
		tmp[cnt] = append(tmp[cnt], str)
	}
	res := make([][]string, 0, len(tmp))
	for _, v := range tmp {
		res = append(res, v)
	}
	return res
}

// @lc code=end

File: C++/49.字母异位词分组.cpp
/*
 * @lc app=leetcode.cn id=49 lang=cpp
 *
 * [49] 字母异位词分组
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<vector<string>> res;
        map<string, int> smap;
        for(auto str : strs) {
            string word = str;
            sort(word.begin(), word.end());
            if(smap.find(word) == smap.end()) {
                vector<string> ivec;
                ivec.push_back(str);
                res.push_back(ivec);
                smap[word] = res.size() - 1;
            } else {
                res[smap[word]].push_back(str);
            }
        }
        return res;
    }
};
// @lc code=end

